src/client/Client.cc

```diff
@@ -11942,11 +11945,16 @@ void Client::refresh_snapdir_attrs(Inode *in, Inode *diri) {
   in->gid = diri->gid;
   in->nlink = 1;
   in->mtime = diri->mtime;
-  in->ctime = diri->ctime;
+  in->change_attr = 0;
```

Comment:
@lxbsz @ajarr - WDYT about this? Using `diri->change_attr` by default can be a bit confusing to users (but `diri->mtime` is used by default)

Reply:
@vshankar IMO it makes no sense to set `in->mtime` to `diri->mtime`, we should always copy both the `in->mtime` and `in->change_attr` from realm. 

Just in case:

`dirA/dirB/dirC/`, if the `snaprealm` is located in `dirA/` and when `refresh_snapdir_attrs()` for `dirC/.snap/` then here it could always be `diri_C->mtime > snapreaml->last_modified` when after the last snapshot was made the `dirC/` was updated frequently.

And at the same time in the MDS side when a snaprealm is first created it's initial `last_modified` should be properly set and shouldn't be `0`.

Reply:
> @vshankar IMO it makes no sense to set `in->mtime` to `diri->mtime`, we should always copy both the `in->mtime` and `in->change_attr` from the snaprealm.

That happens when `realm->change_attr` is non-zero. What should `in->mtime` be (for snapdir) otherwise?

> 
> Just in case:
> 
> `dirA/dirB/dirC/`, if the `snaprealm` is located in `dirA/` and when `refresh_snapdir_attrs()` for `dirC/.snap/` then here it could always be `diri_C->mtime > snapreaml->last_modified` when after the last snapshot was made the `dirC/` was updated frequently.

Wouldn't there be a snalrealm for dirC in this case?

> 
> And at the same time in the MDS side when a snaprealm is first created it's initial `last_modified` should be properly set and shouldn't be `0`.

Its done when a snap is created, deleted or renamed. `last_modified` would be correctly initialized when decoding `sr_t`, isn't it?

Reply:
OK - I think I get your point @lxbsz. `refresh_snapdir_attrs()` can be called for a subdir for which there is no realm (yet). What should ->mtime and ->change_attr be for the snapdir in that case?

Reply:
> OK - I think I get your point @lxbsz. `refresh_snapdir_attrs()` can be called for a subdir for which there is no realm (yet). 

If I didn't miss something important, from my reading from the code:

If we didn't make any snapshot under `dirC/.snap/` directly, so for both `dirB/` and `dirC/` will point to the same snaprealm belonged to `dirA/`. Only after we make a snapshot in `dirC/.snap/` will it create a new dedicated snaprealm for `dirC/`.


> What should ->mtime and ->change_attr be for the snapdir in that case (which is what my question was initially for this code block).

IMO it should be the snaprealm's `last_modified` and `change_attr`. Such as for `dirC/` we will always get the `dirA/`'s snaprealm. That means for `dirB/` and `dirC/` should use the `dirA/`'s snaprealm's `last_modified` and `change_attr`.

But if we couldn't find any, (I still have no idea in which case we couldn't find one, maybe for a fresh fs if we didn't make any snapshot then even the '/' dir won't have a snaprealm ? IMO it should point to the global snaprealm in this case.) we can use the global snaprealm's, which should always exist.

Reply:
> > OK - I think I get your point @lxbsz. `refresh_snapdir_attrs()` can be called for a subdir for which there is no realm (yet).
> 
> If I didn't miss something important, from my reading from the code:
> 
> If we didn't make any snapshot under `dirC/.snap/` directly, so for both `dirB/` and `dirC/` will point to the same snaprealm belonged to `dirA/`. Only after we make a snapshot in `dirC/.snap/` will it create a new dedicated snaprealm for `dirC/`.

Inodes for `dirB` and `dirC` "join" snaprealm for `dirA` - what I'm seeing from the logs is that the snap trace is not sent back to the client for `dirB` and `dirC`. I need to check the code where this is done (if that's the approach).

> 
> > What should ->mtime and ->change_attr be for the snapdir in that case (which is what my question was initially for this code block).
> 
> IMO it should be the snaprealm's `last_modified` and `change_attr`. Such as for `dirC/` we will always get the `dirA/`'s snaprealm. That means for `dirB/` and `dirC/` should use the `dirA/`'s snaprealm's `last_modified` and `change_attr`.

If we do this:

```
dirC_snapdir->change_attr = dirA_realm->change_attr;
dirC_snapdir->mtime = dirA_realm->mtime;
```

Then when snaprealm for `dirC` is created (when taking a snap), its `change_attr` starts from 0. Either we ensure that `change_attr` is always monotonically increasing or (from what I'm seeing) snap realm for `dirC` is nullptr in the client.

> 
> But if we couldn't find any, (I still have no idea in which case we couldn't find one, maybe for a fresh fs if we didn't make any snapshot then even the '/' dir won't have a snaprealm ? IMO it should point to the global snaprealm in this case.) we can use the global snaprealm's, which should always exist.

There would at least exist a global snap realm.

Reply:
> > > OK - I think I get your point @lxbsz. `refresh_snapdir_attrs()` can be called for a subdir for which there is no realm (yet).
> > 
> > 
> > If I didn't miss something important, from my reading from the code:
> > If we didn't make any snapshot under `dirC/.snap/` directly, so for both `dirB/` and `dirC/` will point to the same snaprealm belonged to `dirA/`. Only after we make a snapshot in `dirC/.snap/` will it create a new dedicated snaprealm for `dirC/`.
> 
> Inodes for `dirB` and `dirC` "join" snaprealm for `dirA` - what I'm seeing from the logs is that the snap trace is not sent back to the client for `dirB` and `dirC`. I need to check the code where this is done (if that's the approach).

If my understanding is correct, only when the `dirB/` and `dirC/` have snapshots will the client get the snap trace of them and create a corresponding snaprealm entities for them. Or they will refer to `dirA/`'s snaprealm entity always.

> 
> > > What should ->mtime and ->change_attr be for the snapdir in that case (which is what my question was initially for this code block).
> > 
> > 
> > IMO it should be the snaprealm's `last_modified` and `change_attr`. Such as for `dirC/` we will always get the `dirA/`'s snaprealm. That means for `dirB/` and `dirC/` should use the `dirA/`'s snaprealm's `last_modified` and `change_attr`.
> 
> If we do this:
> 
> ```
> dirC_snapdir->change_attr = dirA_realm->change_attr;
> dirC_snapdir->mtime = dirA_realm->mtime;
> ```
> 
> Then when snaprealm for `dirC` is created (when taking a snap), its `change_attr` starts from 0. Either we ensure that `change_attr` is always monotonically increasing or (from what I'm seeing) snap realm for `dirC` is nullptr in the client.

Yeah, this is what I mentioned in https://github.com/ceph/ceph/pull/48382#discussion_r1011729981 we should always make sure it should be properly initialized. IMO a new snaprealm should always inherit `change_attr` from the parent snaprealm.

Reply:
> If my understanding is correct, only when the `dirB/` and `dirC/` have snapshots will the client get the snap trace of them and create a corresponding snaprealm entities for them. Or they will refer to `dirA/`'s snaprealm entity always.

The userspace client has

```
ceph::unordered_map<inodeno_t,SnapRealm*> snap_realms;
```

The snaprealm itself has reference to its parent (`pparent`) snaprealm and the parent has a list of children snaprealms (`pchildren`). E.g.: For `dirA/dirB/dirC`, if `dirA` had a snaprealm and a snap is created in `dirC`, `dirC_realm::pparent` would be snaprealm of `dirA`. `dirB` snaprealm is still nullptr AFAICT.

Reply:
OK - so `in->snaprealm` can point to an ancestors snaprealm and `snap_realms` might not have an entry for this inode. I guess this is what you mean @lxbsz.

Reply:
> > If my understanding is correct, only when the `dirB/` and `dirC/` have snapshots will the client get the snap trace of them and create a corresponding snaprealm entities for them. Or they will refer to `dirA/`'s snaprealm entity always.
> 
> The userspace client has
> 
> ```
> ceph::unordered_map<inodeno_t,SnapRealm*> snap_realms;
> ```
> 
> The snaprealm itself has reference to its parent (`pparent`) snaprealm and the parent has a list of children snaprealms (`pchildren`). E.g.: For `dirA/dirB/dirC`, if `dirA` had a snaprealm and a snap is created in `dirC`, `dirC_realm::pparent` would be snaprealm of `dirA`.

Yeah, correct. The hierarchy will be:

                        dirA/ (snaprealmA)
                           /            \
                      dirB/         dirD/
                         \
                       dirC/ (snaprealmC)

>  `dirB` snaprealm is still nullptr AFAICT.

Yeah. Shouldn't `dirB`'s `in->snaprealm` will always point to `snaprealmA` once the `in->is_any_caps()` is true ? Please see `Client::add_update_cap()`.
IMO for all the `dirA`, `dirB` and `dirD` snapdirs they should always use the `snaprealmA`'s `change_attr` and `last_modified`.  This is what the PR #47879 should work as, but it seems there is a little buggy and I need to test it more to confirm it works as expected as the kernel corresponding patch.

Reply:
Yeh. the `snap_realms` maps confused me a bit. `in->snaprealms` points to the appropriate snaprealm. We'd need to start from global (or root) snaprealm `last_modified`/`change_attr` and then adjust these accordingly (on created snaprealms) down the heirarchy.

Reply:
Sounds good :-)